### PR TITLE
Amazon EMR on Amazon EKS

### LINKED ISSUE #17175: Operators for orchestrating EMR EKS Jobs
**Description**

[EMR in EKS](https://aws.amazon.com/emr/features/eks/) is a new feature by AWS. For teams that heavily uses EKS for their workloads, it would make sense to submit Jobs (Spark Jobs) via [EMR Containers](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/emr-containers.html) and leverage execution with EKS.
This issue is request to add operators that would enable starting jobs in EMR EKS.

**Use case / motivation**

Enable EMR jobs on EKS orchestration through Airflow. This would enable airflow users to create dags that submit jobs to EMR that needs to use EKS for execution power.

**Are you willing to submit a PR?**

I have already created the operators and hooks to enable this. PR is on its way.

**Related Issues**

N/A

### PR DESCRIPTION
**Pull request by wanderijames**
Adding operators for interacting with EMR on EKS as per https://github.com/apache/airflow/issues/17175

- Add operator for starting job
- Add operator for canceling a job
- Add operator for checking job state

closes: https://github.com/apache/airflow/issues/17175
related: https://github.com/apache/airflow/issues/17175

### CODE REVIEW

**Comment by mik-laj** [comment]
We currently have one operator that allows us to run Spark job on Kubernetes. It works with both EKS and GCP as well as any other Kubernetes platform. - [SparkKubernetesOperator](https://github.com/apache/airflow/blob/d72b363929c86eb03fc9583002459bd10bc7eaeb/airflow/providers/cncf/kubernetes/operators/spark_kubernetes.py#L24). Why would anyone use this operator instead of the generic operator for Kubernetes?

**Comment by wanderijames** [comment]
> We currently have one operator that allows us to run Spark job on Kubernetes. It works with both EKS and GCP as well as any other Kubernetes platform. - [SparkKubernetesOperator](https://github.com/apache/airflow/blob/d72b363929c86eb03fc9583002459bd10bc7eaeb/airflow/providers/cncf/kubernetes/operators/spark_kubernetes.py#L24). Why would anyone use this operator instead of the generic operator for Kubernetes?

Hey @mik-laj, I am aware of apache spark and livy operator and also EMR operator. However, EMR on EKS works differently because EMR launches virtual cluster in your EKS. The pods (spark master and executors) launched are ephemeral, only existing when a start job is invoked. For more information, kindly visit https://aws.amazon.com/emr/features/eks/

File: airflow/providers/amazon/aws/operators/emr_containers_get_job_state.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+# specific language governing permissions and limitations
+# under the License.
+
+from typing import Any, Dict
+
+from airflow.models import BaseOperator
+from airflow.providers.amazon.aws.hooks.emr_containers import EmrContainersHook
+
+
+class EmrContainersGetJobStateOperator(BaseOperator):
```
Comment (dacort) [inline airflow/providers/amazon/aws/operators/emr_containers_get_job_state.py:25]: I'm curious what the use case is for an operator that just returns the job state? 

Apologies if this is a silly question - still getting up to speed on typical Airflow patterns.

File: airflow/providers/amazon/aws/operators/emr_containers_get_job_state.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    """Operator to get a job status.
+
+    A job run is a unit of work, such as a Spark jar, PySpark script, or SparkSQL query,
+    that you submit to Amazon EMR on EKS.
+
+    A job is in PENDING, SUBMITTED, RUNNING, FAILED, CANCELLED or CANCEL state
```
Comment (dacort) [inline airflow/providers/amazon/aws/operators/emr_containers_get_job_state.py]:
```suggestion
    A job is in PENDING, SUBMITTED, RUNNING, FAILED, CANCELLED, CANCEL_PENDING, or COMPLETED state
```

Per https://docs.aws.amazon.com/emr-on-eks/latest/APIReference/API_JobRun.html

File: airflow/providers/amazon/aws/operators/emr_containers_start_job.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        super().__init__(**kwargs)
+        self.start_job_params = dict(
+            cluster_id=cluster_id,
+            execution_role_arn=execution_role_arn,
+            emr_release_label=emr_release_label,
+            job_driver=job_driver,
+            configuration_overrides=configuration_overrides,
+            tags=tags,
+            name=name,
+            client_token=client_token,
```
Comment (dacort) [inline airflow/providers/amazon/aws/operators/emr_containers_start_job.py:76]: I don't see you defining `client_token` anywhere. Have you tested this in a live environment to validate that it works?

**Review by dacort — CHANGES_REQUESTED** [review]
Added a few comments/questions. 

Realizing there is https://github.com/apache/airflow/pull/16766 already submitted, the big difference between the two seems to be the addition of a couple different operators and the default behavior in #16766 being that the Operator waits for the job to complete.

I haven't done a thorough review of this, but I think (depending on if they're needed) we should try to merge 1 or more of the additional operators into my PR. I'm just not familiar enough with real-world DAGs to know to which degree the additional operators are necessary. If you have any feedback on that, would much appreciate it!

File: airflow/providers/amazon/aws/operators/emr_containers_start_job.py (hunk 3 above)
Comment (wanderijames) [reply airflow/providers/amazon/aws/operators/emr_containers_start_job.py:76]: It is being used in line 89;
```python
response = emr_containers.start_job(**self.start_job_params)
```

File: airflow/providers/amazon/aws/operators/emr_containers_get_job_state.py (hunk 1 above)
Comment (wanderijames) [reply airflow/providers/amazon/aws/operators/emr_containers_get_job_state.py:25]: Now that you asked me that question, I don't think this has a straight forward use case as an operator.

File: airflow/providers/amazon/aws/operators/emr_containers_start_job.py (hunk 3 above)
Comment (dacort) [reply airflow/providers/amazon/aws/operators/emr_containers_start_job.py:76]: Right, but it's not initialized anywhere. The default is `None`, which means if the client doesn't pass in a `client_token`, a new job won't get created after the first one. I ran into a similar issue and had to generate a UUID as the default:
```
client_token or str(uuid4())
```

**Comment by dacort** [comment]
Here's my current thinking on this - I like the addition of the `Start` and `Cancel` operators. I think we should try to get #16766 merged and then you can 1. rebase off that commit and 2. add documentation to the `emr_eks.rst` file on how to use the two additional operators. 

Thoughts?

File: airflow/providers/amazon/aws/operators/emr_containers_start_job.py (hunk 3 above)
Comment (wanderijames) [reply airflow/providers/amazon/aws/operators/emr_containers_start_job.py:76]: Thanks @dacort for this.

**Comment by wanderijames** [comment]
> Here's my current thinking on this - I like the addition of the `Start` and `Cancel` operators. I think we should try to get #16766 merged and then you can 1. rebase off that commit and 2. add documentation to the `emr_eks.rst` file on how to use the two additional operators.
> 
> Thoughts?

I agree @dacort , let's do that.

**Comment by eladkal** [comment]
@wanderijames https://github.com/apache/airflow/pull/16766 was merged. Would you like to proceed?

**Pull request closed without merging**